RavenDB-14738 counters indexing

ppekrol opened the pull request.

ppekrol committed:
RavenDB-14738 started working on counters indexing

ppekrol committed:
RavenDB-14738 more work on counters indexing

ppekrol committed:
RavenDB-14738 small refactoring

ppekrol committed:
RavenDB-14738 basic indexing test now works

ppekrol committed:
RavenDB-14738 added document references handling

ppekrol committed:
RavenDB-14738 support for map-reduce

ppekrol committed:
RavenDB-14738 minor adjustments

ppekrol committed:
RavenDB-14738 added test for map-reduce with load

ayende commented on src/Raven.Server/Documents/Indexes/Static/Counters/MapCountersIndex.cs:
```diff
@@ -0,0 +1,248 @@
+                return;
+            }
+
+            if (_referencedCollections.Count > 0 && _referencedCollections.Contains(change.CollectionName))
```
```suggestion
            if ( _referencedCollections.Contains(change.CollectionName))
```

Do the same thing, no?

ppekrol replied:
Yeah.

ayende commented on src/Raven.Server/Documents/Indexes/Static/Counters/MapCountersIndex.cs:
```diff
@@ -0,0 +1,248 @@
+        {
+            if (change.Type == DocumentChangeTypes.Delete && (HandleAllDocs || Collections.Contains(change.CollectionName)))
+            {
+                // in counters we need to subscribe only to deletions of source documents
```
Why? If the referenced document was deleted, shouldn't we reindex?

ppekrol replied:
We are, it is handled below.

ayende replied:
So why do we need this code?

ppekrol replied:
This is for source document deletion, not referenced. E.g. I'm indexing counters from companies/1. We need to trigger the indexing if it was deleted.

ayende commented on test/FastTests/Client/Indexing/Counters/BasicCountersIndexes.cs:
```diff
@@ -0,0 +1,741 @@
+                {
+                    Name = indexName,
+                    Maps = {
+                    "from counter in counters.Users.HeartRate " +
```
How can I index all counters in a collection? `counters.Users` would work?

ppekrol replied:
Yes, it will work the same way as for timeseries, but I haven't implemented that yet. Going one feature by one.

ppekrol committed:
RavenDB-14738 do not need to check count, hashset is optimized enough

ppekrol committed:
RavenDB-14738 fixing compilation

ppekrol committed:
RavenDB-14738 counters need to be case insensitive

ppekrol committed:
RavenDB-14738 support for updating index

ppekrol committed:
RavenDB-14738 added update test for map-reduce index

ppekrol committed:
RavenDB-14738 added support for indexing all counters from a collection

ppekrol committed:
RavenDB-14738 fixed counter index persistence

ppekrol committed:
RavenDB-14738 started working on AbstractCountersIndexCreationTask

ppekrol committed:
RavenDB-14738 fixed index type detection

ppekrol committed:
RavenDB-14738 added test for load

ppekrol committed:
RavenDB-14738 added map-reduce tests

ppekrol committed:
RavenDB-14738 added tests for mapping all counters and all docs

ppekrol committed:
RavenDB-14738
- added multi-map test and query test
- fixed document id extraction, we need to recalculate escape positions when slicing

ppekrol committed:
RavenDB-14738 added map-reduce query test

ppekrol committed:
RavenDB-14738
- properly handling counter references (we need to store them by key to not remove excessive records e.g. when we have multi-map on the same collection, but different counter name)
- compare exchange handling for counters

ppekrol committed:
RavenDB-14738 added compare exchange query test

ppekrol committed:
RavenDB-14738 fixing comments

ppekrol commented on src/Raven.Server/Documents/Indexes/Static/Counters/CounterItemFilterBehavior.cs:
```diff
@@ -0,0 +1,19 @@
+{
+    public class CounterItemFilterBehavior : IIndexItemFilterBehavior
+    {
+        private readonly HashSet<LazyStringValue> _seenIds = new HashSet<LazyStringValue>();
```
TODO:
- this is just an optimization, maybe we do not need to keep everything, just last X items
- check if we are not disposing LSV

ppekrol committed:
RavenDB-14738 we need to use strings because keys are being disposed

arekpalinski commented on src/Raven.Client/Documents/Indexes/IndexDefinitionHelper.cs:
```diff
@@ -239,22 +242,30 @@ internal static IndexSourceType DetectStaticIndexSourceType(string map)
+
+            if (map.StartsWith("counters", StringComparison.OrdinalIgnoreCase))
+            {
+                return IndexSourceType.TimeSeries;
```
Invalid source type - it should be `IndexSourceType.Counters`

ppekrol replied:
done

arekpalinski commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -1796,5 +1805,239 @@ public long GetLastCounterEtag(DocumentsOperationContext context, string collect
+
+                    if (valuesData.TryGet(Values, out BlittableJsonReaderObject values))
+                    {
+                        size /= values.Count; // just 'estimating'
```
Whats the purpose of size? What do we use it for?

ppekrol replied:
For stats, we are using this to record the amount of data that was processed

ayende replied:
Can Count be 0 here?

ppekrol replied:
spoke with @aviv86 it is not possible

arekpalinski commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -250,6 +251,9 @@ protected Index(IndexType type, IndexSourceType sourceType, IndexDefinitionBase
                     _itemType = "time series item";
                     break;
+                case IndexSourceType.Counters:
+                    _itemType = "counter item";
```
Just "counter" maybe?

ppekrol replied:
done

arekpalinski commented on src/Raven.Server/Documents/Indexes/Static/Counters/CounterItemFilterBehavior.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Raven.Server.Documents.Indexes.Static.Counters
+{
+    public class CounterItemFilterBehavior : IIndexItemFilterBehavior
```
Can you please put some comment explaining why do we have it? Something that it's related to how counters are stored, that we have counter groups and implications for indexing because of that

ppekrol replied:
done

arekpalinski commented on src/Raven.Server/Documents/Indexes/Static/Counters/DynamicCounterEntry.cs:
```diff
@@ -0,0 +1,63 @@
+                .DocumentDatabase
+                .DocumentsStorage
+                .CountersStorage
+                .GetCounterValue(current.QueryContext.Documents, _counterItemMetadata.DocumentId, _counterItemMetadata.CounterName);
```
`.QueryContext`? We call it during indexing not when querying it - please check it

ppekrol replied:
It is fine, in this case the name does not match very well, but indexing batch also needs to open that context, because we might be indexing compare exchange

arekpalinski commented on src/Raven.Server/Documents/Indexes/Static/Counters/MapCountersIndex.cs:
```diff
@@ -0,0 +1,248 @@
+
+        public override (ICollection<string> Static, ICollection<string> Dynamic) GetEntriesFields()
+        {
+            throw new NotImplementedException();
```
TODO ?

ppekrol replied:
done

ppekrol committed:
RavenDB-14738 adjustments

ppekrol committed:
RavenDB-14738 fixing test

ppekrol committed:
RavenDB-14738 fixed compare exchange reference handling in map-reduce counters index

ppekrol committed:
RavenDB-14738 fixing compilation
